### DISCUSSION TITLE
Ionisation doesn't work

### DISCUSSION THREAD

**Discussion by hslewk** [epochpic/epoch#564] [General]
Dear friends,

 I want to use an electron beam to ionize hydrogen, the code can run normally but the NOTE appears at the end of the run. The NOTE is: The following floating-point exceptions are signalling: IEEE_INVALID_FLAG IEEE_DIVIDE_BY_ZERO
<img width="588" alt="93ddb38c34314259d54d19559010727" src="https://github.com/Warwick-Plasma/epoch/assets/102033233/f03c08dd-0b98-4033-9f16-8d6714a108ea">
Also,  hydrogen does not be ionized. Although there are "hydrogen1" and "electron" in output file, the number density of them are all 0.
Here is my input file, and the code version I used is 4.19.0.

 begin:constant
  ele_dens = 1.e21
  bac_dens = 1.e24
  lambda0 = 1.e-6   
  T0 = lambda0/c  #T0=10/3 fs
  v = 0.29979
end:constant

begin:control
  nx = 400
  ny = 400
  x_min = 0 * micron
  x_max = 40 * micron
  y_min = -20 * micron
  y_max = 20* micro  
  t_end = 160 * femto 
  use_multiphoton = F
  use_bsi = F
  field_ionisation=T
  physics_table_location = /media/lt/2/epoch-4.19.0/epoch2d/src/physics_packages/TABLES
  stdout_frequency = 10
 end:control

 
begin:boundaries
  bc_x_min = simple_outflow
  bc_x_max = simple_outflow
  bc_y_min = open
  bc_y_max = open
end:boundaries

begin:injector
boundary = x_min
species = electron1
t_start = 0
t_end  =  4*T0
nparticles_per_cell=5
number_density= if ((y gt -15*micron) and (y lt 15*micron), ele_dens, 0)
temp_ev=0
drift_x = 5.3442e-20
end:injector

begin:species
  name = electron1
  charge = -1.0
  mass = 1.0
  bc_x_min = simple_outflow
  bc_x_max = simple_outflow
  nparticles_per_cell=6
end:species

######################################################

begin:collisions
use_collisions = T
use_nanbu = T
coulomb_log = auto
collide = all
collisional_ionisation = T
ci_n_step = 3
end:collisions

######################################################

begin:species
name = hydrogen
charge = 0
atomic_no = 1
ionise = T
ionise_limit = 1
mass = 1836.2*1
number_density = if ((x gt 4*micron) and (x lt 40*micron), bac_dens, 0)
ionisation_electron_species = electron
nparticles_per_cell = 3
end:species

begin:species
charge = -1.0
mass = 1.0
name = electron
nparticles_per_cell = 3
end:species

begin:window
move_window =F
window_v_x = 0.9995*c
window_start_time = 60*T0
bc_x_min_after_move = simple_outflow   
bc_x_max_after_move = simple_outflow   
end:window

##############################################################

begin:output
name = out1
file_prefix = field
dt_snapshot = 20*femto  
grid = always + single
ex = always + single
ey = always + single
bx = always + single
by = always + single
end:output

begin:output
name = out2
file_prefix = momentum
dt_snapshot = 20*femto  
grid = always + single
average_px = always+species+no_sum+single
average_py = always+species+no_sum+single
end:output

begin:output
name = out3
file_prefix = number_density
dt_snapshot = 10*femto  
number_density = always + single + species + no_sum
end:output

**Comment by Status-Mirror** [comment]
Hi @hslewk,

I'm not sure what the floating point exceptions are describing, but I can see why you have no ionisation here.

You include field ionisation, but there is no laser in your simulation, so I doubt you could reach the fields required to ionise hydrogen.

You also include collisional ionisation, but you have no electron species which can trigger it. By including `ionisation_electron_species = electron`, the code understands that `electron` is an electron species which can trigger collisional ionisation, but this species starts with no particles in it. You use `nparticles_per_cell = 3` in the `electron` species block, but you don't specify a number density for the species, so nothing is loaded. You _do_ have electrons in the `electron1` species, but you haven't told EPOCH that this is an electron species. This can be done by inserting `identify:electron` into the `electron1` species block. See the Ionisation section in the [species block](https://epochpic.github.io/documentation/input_deck/input_deck_species.html) documentation.

Hope this helps,
Stuart

**Comment by hslewk** [comment]
Thank you @Status-Mirror 

But  I try to inset ### identify:electron### into the electron1 species block, and the hydrogen still dose not be ionized.

begin:species
name = electron1
charge = -1.0
mass = 1.0
bc_x_min = simple_outflow
bc_x_max = simple_outflow
identify:electron
nparticles_per_cell=6
end:species

**Comment by hslewk** [comment]
And I tried using laser instead of injecting electrons, and still no ionization occurred.

begin:control
  nx = 400
  ny = 400
  x_min = 0 * micron
  x_max = 40 * micron
  y_min = -20 * micron
  y_max = 20* micro  
  t_end = 160 * femto 
  use_multiphoton = T
  use_bsi = F
  field_ionisation=T
  physics_table_location = /media/lt/2/epoch-4.19.0/epoch2d/src/physics_packages/TABLES
  stdout_frequency = 10
 end:control

begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_laser
  bc_y_min = open
  bc_y_max = open
end:boundaries

begin:laser
 boundary = x_min
 intensity_w_cm2 = 1.e23
 lambda = 0.8*micron
 pol_angle = pi/2
 profile = gauss(y,0*micron,14*micron)
 t_profile = gauss(t,35*femto,50*femto)
 phase = 0
 end:laser

**Comment by Status-Mirror** [comment]
Hi @hslewk,

This time I've actually run your simulation, so hopefully I can be a little more helpful.

I don't think you expect to see any ionisation in your simulation. The collisional ionisation package in EPOCH (as discussed in [my paper](https://pubs.aip.org/aip/pop/article/29/12/123907/2843764/Improvements-to-collisional-ionization-models-for)) uses the MBELL model. The cross section of collisional ionisation in hydrogen as a function of incident electron kinetic energy is given by the below figure:

![H_plot](https://github.com/Warwick-Plasma/epoch/assets/38282483/62436acf-b53f-4e8b-86f8-cabd47087b74)

or about $\sigma = 2.3\times 10^{-23}m^2$ for your 100 MeV electrons. These travel through around 36 $\mu m$ of hydrogen at $10^{24} m^{-3}$ density, so each electron is expected to produce $N_\delta = n_i \sigma L \approx 0.0008$ hydrogen-1 ions. You inject a total of $1.2\times 10^{11}$ electrons, and so we expect around $9.6\times 10^6$ ions to be produced. Each hydrogen macro-particle in your simulation represents around $3.3\times 10^9$ real particles, so it is quite unlikely that you would ionise even one of these macro-particles, since a single ionisation would create 300x more ions than I would expect for the full simulation!

If you really want to see ionisation in this system, you would need about 1000 macro-hydrogen per cell, which would be expensive to run.

To ensure ionisation is still working, I increased the electron bunch density to $10^{25}m^{-3}$, and saw lots of ionisation.

Hope this helps,
Stuart

**Reply by hslewk** [reply]
Thank you @Status-Mirror 
Your advice has helped me a lot. Thank you for your help!

**Reply by hslewk** [reply]
Dear @Status-Mirror 
 I want to know how to get the cross section of collisional ionisation in hydrogen or oxygen like the one you showed.

Thank you.

**Reply by Status-Mirror** [reply]
EPOCH uses one of two methods for calculating collisional ionisation cross sections, depending on the atomic number of the target.

For details on the EPOCH method, see Section III.B of [Morris et al](https://pubs.aip.org/aip/pop/article/29/12/123907/2843764/Improvements-to-collisional-ionization-models-for), and the references therein. For hydrogen and oxygen, you'll be using MBELL instead of RBEB.

It can take a bit of time to get these calculations set up, but once you have them, it's pretty quick to go from one material to the other.

Cheers,
Stuart

**Comment by DanRRRR** [comment]
At the end can anyone of you guys post the final and complete input.deck demo example which will finally work with ionization and OFI?

**Reply by hslewk** [reply]
Need to increase the number density of the injected electrons. 

Sorry, I'm a beginner, what is OFI?

**Reply by DanRRRR** [reply]
Optical Field Ionization
